built_value/test/standard_json_plugin_test.dart

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2016, Google Inc. Please see the AUTHORS file for details.
+// All rights reserved. Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+import 'package:built_collection/built_collection.dart';
+import 'package:built_value/standard_json_plugin.dart';
+import 'package:built_value/serializer.dart';
+import 'package:test/test.dart';
+
+void main() {
+  final serializers = (new Serializers().toBuilder()
+        ..addPlugin(new StandardJsonPlugin())
+        ..addBuilderFactory(
+            const FullType(BuiltList, const [const FullType(int)]),
+            () => new ListBuilder<int>())
+        ..addBuilderFactory(
+            const FullType(BuiltList, const [
+              const FullType(BuiltList, const [const FullType(int)])
+            ]),
+            () => new ListBuilder<BuiltList<int>>())
+        ..addBuilderFactory(
+            const FullType(
+                BuiltMap, const [const FullType(int), const FullType(String)]),
+            () => new MapBuilder<int, String>())
+        ..addBuilderFactory(
+            const FullType(BuiltMap,
+                const [const FullType(String), const FullType(String)]),
+            () => new MapBuilder<String, String>())
+        ..addBuilderFactory(
+            const FullType(BuiltMap, const [
+              const FullType(BuiltMap,
+                  const [const FullType(int), const FullType(String)]),
+              const FullType(
+                  BuiltMap, const [const FullType(int), const FullType(String)])
+            ]),
+            () =>
+                new MapBuilder<BuiltMap<int, String>, BuiltMap<int, String>>()))
+      .build();
+
+  group('Serializers with StandardJsonPlugin', () {
+    test('throws on serialize if specifiedType is missing', () {
+      expect(() => serializers.serialize(1), throws);
```

Comment:
The default serialization can handle serializing without specifiedType; this means it has to write the type name on the wire so that it can be deserialized correctly.

But the serialization added with this pull request always requires a specifiedType because it never writes type names on the wire. This means that, as you point out, the interface for 'serializers' is not ideal for this case. It would make sense to wrap it.

Someone suggested adding a way to do map-based serialization that also writes types on the wire; I expect to add this as another plugin you can choose. Then you won't always need specifiedType, but you'll need something that can understand the type names.